bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCache.java

```diff
@@ -0,0 +1,89 @@
+/**
+ * Copyright (c) 2014-2017 by the respective copyright holders.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.core.cache;
+
+import java.io.IOException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This is a simple cache implementation of a single key-value-pair.
+ *
+ * There must be provided an action in order to retrieve/calculate the value for any given key. This action will be
+ * called only if the answer from the last calculation is not valid anymore, i.e. if the key is different or the last
+ * result is expired.
+ *
+ * @author Simon Kaufmann - Initial contribution and API.
+ * @author Christoph Weitkamp - Moved ExpiringCache from Yahoo Weather Binding to Eclipse SmartHome Core.
+ *
+ * @param <Q> the type of the key
+ * @param <V> the type of the value
+ */
+public class ExpiringCache<Q, V> {
+
+    private final Logger logger = LoggerFactory.getLogger(ExpiringCache.class);
+
+    private final int expiry;
+    private final LoadAction<Q, V> action;
+    private long lastUpdate;
+    private Q lastQuery;
+    private V lastValue;
+
+    public interface LoadAction<Q, V> {
+        V load(Q key) throws IOException;
+    }
+
+    /**
+     * Create a new instance.
+     *
+     * @param expiry duration in milliseconds for how long a result stays valid
+     * @param action
+     */
+    public ExpiringCache(int expiry, LoadAction<Q, V> action) {
+        if (action == null) {
+            throw new IllegalArgumentException(
+                    "ExpiringCache cannot be created as action is not an instance of LoadAction.");
+        }
+
+        this.expiry = expiry;
+        this.action = action;
+    }
+
+    /**
+     * Get the value for the given key - possibly from the cache, if it is still valid.
+     *
+     * @param key the input parameter
+     * @return the value for the given key
+     * @throws IOException
+     */
+    public synchronized V get(Q key) throws IOException {
+        if (key == null) {
+            throw new IllegalArgumentException("Value cannot be retrieved/calculated as key is null.");
+        }
+
+        if (lastQuery == null || !lastQuery.equals(key) || isExpired()) {
+            lastValue = action.load(key);
+            lastQuery = key;
+            lastUpdate = System.currentTimeMillis();
```

Comment:
How about [using nanoTime](http://efectivejava.blogspot.de/2013/09/why-to-use-systemnanotime-instead-of.html) to make sure we are not impacted by adjusted system clocks?